File: tsc/src/dsa/two-sum/two-sum.test.ts
import { describe, expect, it } from 'vitest'
import { twoSum } from './two-sum'

describe('twoSum', () => {
  it('should return the indices of the two numbers that add up to the target 0,1', () => {
    const nums = [2, 7, 11, 15]
    const target = 9
    const result = twoSum(nums, target)
    expect(result).toEqual([0, 1])
  })

  it('should return null if no two numbers add up to the target', () => {
    const nums = [2, 7, 11, 15]
    const target = 10
    const result = twoSum(nums, target)
    expect(result).toBeNull()
  })

  it('should return the indices of the two numbers that add up to the target 1,2', () => {
    const nums = [9, 2, 5, 6]
    const target = 7

    const result = twoSum(nums, target)
    expect(result).toEqual([1, 2])
  })
})

File: go/test/two-sum/two-sum_test.go
package twosum_test

import (
	"testing"

	twosum "github.com/gusramirez-aplazo/dsa/go/internal/two-sum"
)

func TestTwoSum(t *testing.T) {
	nums := []int{9, 2, 5, 6}
	target := 7

	result := twosum.Exec(nums, target)

	if result == nil {
		t.Errorf("Expected result to be non-nil")
	}

	if len(result) != 2 {
		t.Errorf("Expected result to have length 2")
	}

	if result[0] != 1 || result[1] != 2 {
		t.Errorf("Expected result to be [1, 2]")
	}
}

func TestTwoSum_WithNullResult(t *testing.T) {
	nums := []int{1, 2, 3, 4, 5}
	target := 10

	result := twosum.Exec(nums, target)

	t.Logf("result: %v", result)

	if result != nil {
		t.Errorf("Expected result to be nil")
	}

	if len(result) != 0 {
		t.Errorf("Expected result to have length 0")
	}
}
